cloud/cluster/subscription.go

```diff
@@ -0,0 +1,153 @@
+package cluster
+
+import (
+	"sort"
+	"time"
+)
+
+// Returns a full list of visible nodes.
+type Fetcher interface {
+	Fetch() ([]Node, error)
+}
+
+// Subscribe subscribes to node updates for the specified service
+func Subscribe(fetcher Fetcher) Subscription {
+	ticker := time.NewTicker(time.Duration(15 * time.Second))
+	ch := MakeDiffSubscription(ticker.C, fetcher)
+	return Subscription{nil, ch, &closer{ticker}}
+}
+
+type closer struct {
+	ticker *time.Ticker
+}
+
+func (c *closer) Close() error {
+	c.ticker.Stop()
+	return nil
+}
+
+func MakeDiffSubscription(ticker <-chan time.Time, fetcher Fetcher) chan []NodeUpdate {
+	c := diffSubscription{
+		fetcher:  fetcher,
+		nodes:    make(map[string]Node),
+		tickCh:   ticker,
+		fetchCh:  nil,
+		outCh:    make(chan []NodeUpdate),
+		outgoing: nil,
+	}
+	go c.loop()
+	return c.outCh
+}
+
+// diffSubscription fetches the list of nodes periodically, diffs the list to determine
+// what has changes, and sends updates.
+type diffSubscription struct {
```

Comment:
I think cluster and subscription are logically different things and it makes some sense to keep them separate. We can discuss code organization offline.